cpp/ql/src/semmle/code/cpp/exprs/Expr.qll

```diff
@@ -453,12 +453,19 @@ class Expr extends StmtParent, @expr {
    * cast from B to C. Only (1) and (2) would be included.
    */
   Expr getExplicitlyConverted() {
-    // result is this or one of its conversions
-    result = this.getConversion*() and
-    // result is not an implicit conversion - it's either the expr or an explicit cast
-    (result = this or not result.(Cast).isImplicit()) and
-    // there is no further explicit conversion after result
-    not exists(Cast other | other = result.getConversion+() and not other.isImplicit())
+    if this.hasConversion()
+    then
+    if this.hasExplicitConversion()
+    then
+     result = this.getConversion().getExplicitlyConverted()
+    else
+      if this.getImplicitlyConverted().hasExplicitConversion()
+      then
+        result = this.getImplicitlyConverted().getConversion().getExplicitlyConverted()
+      else
+        result = this
+    else
+    result = this
```

Comment:
I think this is both clever and correct, but it's hard to understand. Some implementation comments would help, along the lines of:

In general, this expression has a chain of 0 or more conversions on it. Each conversion is either implicit or explicit. If the next conversion is explicit, do a recursive call on that. If instead the next conversion starts a sequence of implicit conversions with an explicit conversion at the end, do a recursive call on the explicit conversion at the end.

Reply:
This seems to have correctness issues (if I've written the diff query correctly): https://lgtm.com/query/4098962143601432083/

Reply:
It looks like `hasExplicitConversion` and `hasImplicitConversion` are defined so that only `Cast`s are considered implicit. That seems like a bug, but it's a very old one.

Reply:
I also realized I can simplify this a bit - if `this.hasExplicitConversion()` holds, then `this = this.getImplicitlyConverted()`, so only one `if` is necessary.